src/simd.rs

```diff
@@ -0,0 +1,198 @@
+//! fully generality of the universal algorithm and are hence able to skip a few
+//! instructions.
+
+#[cfg(all(target_arch = "x86_64", feature="simd"))]
+use core::arch::x86_64::*;
+
+pub(crate) enum LoopInstruction<T> {
+    /// Continue looking for more special bytes, but skip next few bytes.
+    ContinueAndSkip(usize),
+    /// Break looping immediately, returning with the given index and value.
+    BreakAtWith(usize, T)
+}
+
+#[allow(overflowing_literals)]
+#[cfg(all(target_arch = "x86_64", feature="simd"))]
+unsafe fn compute_mask(bytes: &[u8], ix: isize) -> i32 {
+    // constants. computed with the below code
+    // ```rust
+    // let chars = [b'\n', b'\r', b'*', b'_', b'~', b'|', b'&', b'\\', b'[', b']', b'<', b'!', b'`'];
+    // let mut lower_bitmap = [0u8; 16];
+    // for &c in &chars {
+    //     lower_bitmap[(c & 0x0f) as usize] |= 1 << (c >> 4);
+    // }
+    // let bitmap = unsafe { _mm_loadu_si128(lower_bitmap.as_ptr() as *const _) };
+    // ```
+    let bitmap = _mm_setr_epi8(
+        64, 4, 0, 0, 0, 0, 4, 0, 0, 0, 5, 32, 168, 33, 128, 32
+    );
+    let bitmask_lookup = _mm_setr_epi8(
+        1, 2, 4, 8, 16, 32, 64, -128,
+        1, 2, 4, 8, 16, 32, 64, -128
+    );
+
+    // actual computation
+    let raw_ptr = bytes.as_ptr().offset(ix) as *const _;
+    let input = _mm_loadu_si128(raw_ptr);
+    let bitset = _mm_shuffle_epi8(bitmap, input);
+    let higher_nibbles = _mm_and_si128(_mm_srli_epi16(input, 4), _mm_set1_epi8(0x0f));
+    let bitmask = _mm_shuffle_epi8(bitmask_lookup, higher_nibbles);
+    let tmp = _mm_and_si128(bitset, bitmask);
+    let result = _mm_cmpeq_epi8(tmp, bitmask);
+    _mm_movemask_epi8(result)
+}
+
+/// Calls callback on bytes. Breaks when callback returns LoopInstruction::BreakAtWith(ix, val). And skips the
+/// number of bytes in callback return value otherwise.
+/// This method returns final index and a possible break value.
+#[cfg(all(target_arch = "x86_64", feature="simd"))]
+pub(crate) fn iterate_special_bytes<F, T>(bytes: &[u8], mut ix: usize, mut callback: F) -> (usize, Option<T>)
+    where F: FnMut(usize) -> LoopInstruction<Option<T>> 
+{
+    let upperbound = bytes.len().saturating_sub(15);
+    
+    while ix < upperbound {
+        let mut mask = unsafe { compute_mask(bytes, ix as isize) };
+        let block_start = ix;
+
+        while mask != 0 {
+            let mask_ix = mask.trailing_zeros() as usize;
+            ix += mask_ix;
+            match callback(ix) {
+                LoopInstruction::ContinueAndSkip(skip) => {
+                    ix += skip + 1;
+                    mask >>= skip + mask_ix + 1;
+                }
+                LoopInstruction::BreakAtWith(ix, val) => return (ix, val),
+            }
+        }
+
+        ix = std::cmp::max(ix, 16 + block_start);
+    }
+
+    let alignment = (bytes.as_ptr() as usize) % 16;
+    let remaining_bytes = bytes.len().saturating_sub(ix);
+
+    if remaining_bytes > 0 {
```

Comment:
Similar concerns as #235; I'd prefer this not to ever read past the slice given. On the other hand, if we can get a determination from an unsafe Rust expert that this is ok, I might soften my stance.